### PR TITLE
fix(tabs): refactor dismissable tab html

### LINKED ISSUE #14820: [a11y]: Tabs [dismissible] - unable to dismiss in iPhone
### Package

@carbon/react

### Browser

Safari & iPhone

### Package version

@carbon/react@1.38.0

### Automated testing tool and ruleset

Manual QA

### Assistive technology

JAWS & VO

### Description
"No way of interaction available for VO users to delete A tab in iPhone.

NOTE: In call its understood there is delete icon which helps the visual users to delete the TAB but its not read by Voiceover."

### Recommended Fix
Provide Delete Name and Button Role for Delete ICON and make it keyboard operable

When the standard interaction is not possible, user is unable to accomplish A task.

### WCAG 2.1 Violation

4.1.2 Name, Role, Value

### PR DESCRIPTION
**Pull request by alisonjoseph**
Closes #14820
Ref #https://github.com/carbon-design-system/carbon/issues/15919

Previously you were only able to dismiss a dismissable tab with the delete key or mouse. This leaves voiceover users on an iphone unaware the dismiss icon even existed. 

This is a complete refactor of the html for dismissable tabs, it changes the icon to a button, and moves it outside of the containing tab button and into the tab order. This allows both keyboard and voiceover users to tab to the icon and dismiss the tabs.

#### Changelog

**Changed**

- Convert the dismiss icon to a button, move outside of the tab button
- Style updates to account for updated html structure
- Update text for screen-readers to say "remove _tab name_ tab" instead of "close tab"
- Updated tests 
- Update focus state after tab is removed

#### Testing / Reviewing

Make sure there are no regressions with any of the tab variants. Be sure to check dismissable in both regular and contained. 

Check that the tab order is correct and that the dismiss icon is read on voice over.

### CODE REVIEW

**Comment by tay1orjones** [comment]
Thanks for digging into this! I think it looks good from a code review standpoint.

The landscape around this is kind of a mess. [This stackoverflow post](https://stackoverflow.com/questions/78071123/aria-tablist-role-that-contains-elements-other-than-tabs) explains it pretty well. There's also a big thread about it that hasn't been resolved in https://github.com/w3c/aria/issues/1440

I think this technically doesn't follow WAI-ARIA 1.2 because [a `tablist` should contain a list `tab` elements](https://www.w3.org/TR/wai-aria/#tablist:~:text=tablist%20role-,A%20list%20of%20tab%20elements%2C%20which%20are%20references%20to%20tabpanel%20elements.,-To%20be%20keyboard). 

<details>
  <summary>It adds an additional "Incomplete" entry from `accessibility-checker` </summary>
  
  image

</details>



@tombrunet when choosing between fixing this or following the spec, I think we should side with the user and fix it regardless of what the spec says, but what do you think?

Also, this contains a change that could break tests or custom functionality in consuming applications because of the shift in the DOM. Despite that, I would support shipping this to all users in our next minor without a feature flag due to the severity of the bug.

**Comment by mbgower** [comment]
So, what we're effectively doing is saying that there is an item in the tab order between the tablist and the tab panel, which is "Dismiss current tab". It happens to be shown inside the tab item, but it is effectively outside the tablist.

I wonder if one approach here is to treat this somewhat like the skiptomain function. Rather than put the focus to the X in the tab item, we make a "Delete tab item" that appears to the right of the tablist after a user tabs out of the tablist. Does that solve anything? It suggests to me it would let us solve the discoverability and actionability requirement AND would not violate ARIA. @tombrunet ?

**Comment by tombrunet** [comment]
@mbgower The largest problem with that approach is that if you change reading modes on the close button, you are out of context. If I go to the close button and then ask to read the previous word, I expect to hear the name of the tab I'm on, not the name of the last tab in the list. 

Looking at some other apps, it does seem that handling of these tab widgets is changing. VS Code has a behavior similar to what Carbon is proposing here (though the close button is only visible for the active tab and on mouse hover). Chrome is actually making every tab widget and close button tabbable.

**Comment by mbgower** [comment]
@alisonjoseph, I've disrupted your fix with my comments, so I wanted to put in an epilogue on this.

I have a caution on the fix which needs to be addressed, as well as some suggestions on future changes, but once that issue is addressed, this looks good to go from an accessibility problem. Details below.

1) ISSUE: It looks like you've implemented this not just on the dismissible tabs, but on ALL the version of Tabs in storybook (default, contained, etc). That shouldn't be happening; it should only exist on the dismissible ones. In the screenshot, you can see that Jaws is focused on a small region to the upper right of the Monitoring tab which has nothing there visually. It is announcing this as "close tab button"
image

2) The implementation on the actual dismissible tab works. In this screen shot, you can see that, after the user presses Tab,  Jaws has properly put focus on the X. Good.
image

3) As I documented in the original issue, ARIA is planning a new attribute aria-actions, which should allow this approach to be possible AND pass current ARIA guidance. @tombrunet  and I have both reviewed your refactor and think what you've proposed not only aligns with some vendor approaches, but also seems to align with this direction of aria. Tom's team will figure out what to do with the Checker rules in regard to this version of the dismissible tab component, since it will currently fail a rule since it does not align with the current aria spec. But it's important to emphasize that it seems to _work_ fine for keyboard and screen reader user. 

4) I would like to suggest that making the message specific to the tab would be useful (i.e., instead of saying "Close tab button" it would say "Close Dashboard tab").

5) I also wonder if "Close" is the correct verb to use in this tooltip/message. "Remove" might make more sense, or even "Delete"? This is question for the content designer, I guess. My slight concern is that since a tab does not normally close (or dismiss, for that matter), a user may be unfamiliar with the operation or notion.

**Comment by alisonjoseph** [comment]
@mbgower thanks for the detailed feedback, I believe I've addressed all of your concerns if you could take another look. 

1. It should no longer be announcing the button all on tab variants.
3. Focus should now go to the next tab after removing a tab, and not to the next close button. 
4. Close buttons should now have a specific message and announce "Remove Dashboard tab" instead of "Close tab"

**Comment by mbgower** [comment]
All these things have been accomplished, nice:

- It should no longer be announcing the close button all on tab variants.
- Focus should now go to the next tab after removing a tab, and not to the next close button.
- Close buttons should now have a specific message and announce "Remove Dashboard tab" instead of "Close tab"

One thing I noticed is that when the last _active_ tab is deleted (Activity tab), things get a little odd, with either the focus going to the disabled 4th Settings delete button, or focus going up the page. 
image
I'd check with @tay1orjones if he remembers the approach we previously discussed about where to place focus in such a situation. I think if it's the last active item and there is a prior item it makes sense to put it there, and that applies if there are no items left at all (i.e., focus would go to Reset). I think that is the basic pattern Josh, Taylor and I discussed many moons ago about what would happen in a table with deletable rows, and the same seems to make sense here as well.

**Comment by alisonjoseph** [comment]
I think this is ready for review now.

@mbgower I've updated it so the previous tab should now take focus after removing a tab.

**Comment by mbgower** [comment]
> @mbgower I've updated it so the previous tab should now take focus after removing a tab. Just wanted to double check that we wouldn't want the active tab to take focus vs the previous one?

I'll have a look, but wanted to clarify that I wasn't saying that each time a tablist item is deleted, the focus should go to the prior tablist item; I think 'next tab item, where available' is a good approach, and supports the APG previously quoted.

I was specifically talking about the situation where the last item in the tablist is deleted  (or the last _active_ item in a tablist). In both cases, I think it probably should go to the prior tablist item.

**Comment by alisonjoseph** [comment]
@mbgower ohhh I completely misread that, let me update this.

**Comment by alisonjoseph** [comment]
@mbgower ok, focus states are updated.

**Comment by mbgower** [comment]
That's working well with JAWS. I also confirmed it is usable and understandable with VO on iOS

**Review by andreancardona — APPROVED** [review]
Also tested with Jaws and looks good to me! 🎉

File: packages/react/src/components/Tabs/Tabs.tsx [hunk 1]
```diff
@@ -849,6 +850,32 @@ const Tab = forwardRef<HTMLElement, TabProps>(function Tab(
   const handleClose = (evt) => {
     evt.stopPropagation();
     onTabCloseRequest?.(index);
+
+    // set focus to corret tab after removing tab
```
Comment (alisonjoseph) [inline packages/react/src/components/Tabs/Tabs.tsx]:
```suggestion
    // set focus after removing tab
```

**Review by tay1orjones — APPROVED** [review]
LGTM!

**Pull request merged**